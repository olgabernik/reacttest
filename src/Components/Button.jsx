import React from 'react';
import PropTypes from 'prop-types';
import styles from '../Modules/Button.module.scss';

function Button({
  name, onClick, isSelected,
}) {
  return (
    <button
      type="button"
      className={isSelected ? styles.buttonClicked : styles.button}
      onClick={onClick}
    >
      {name}
    </button>
  );
}

Button.propTypes = {
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
};

Button.defaultProps = {
  isSelected: false,
};

export default Button;
