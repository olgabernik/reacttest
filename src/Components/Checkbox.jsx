import React from 'react';
import PropTypes from 'prop-types';
import styles from '../Modules/Checkbox.module.scss';

function Checkbox({ isCompleted = false, onCheckTodo, todoid }) {
  return (
    <input
      type="checkbox"
      checked={isCompleted}
      onChange={() => onCheckTodo(todoid)}
      className={styles.roundCheckbox}
    />
  );
}

Checkbox.propTypes = {
  isCompleted: PropTypes.bool.isRequired,
  onCheckTodo: PropTypes.func.isRequired,
  todoid: PropTypes.string.isRequired,
};

export default Checkbox;
