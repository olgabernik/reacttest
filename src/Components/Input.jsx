import React from 'react';
import PropTypes from 'prop-types';
import styles from '../Modules/Input.module.scss';

function Input({
  inputValue, handleInputChange, placeholder, maxLength,
}) {
  return (
    <input
      type="text"
      value={inputValue}
      className={styles.input}
      onChange={handleInputChange}
      maxLength={maxLength}
      placeholder={placeholder}
      data-testid="input-new-todo"
    />
  );
}

Input.propTypes = {
  inputValue: PropTypes.string.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired,
  maxLength: PropTypes.string.isRequired,
};

export default Input;
