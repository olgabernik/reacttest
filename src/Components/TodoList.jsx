import React from "react";
import TodoItem from "./TodoItem";
import styles from "../Modules/TodoList.module.scss";

function TodoList({todos, onDeleteTodo, onCheckTodo}) {
  return (
    <ul className={styles.todoList}>
      {
        todos.map((todo) => (
        <TodoItem key ={todo.id} todo= {todo} onDeleteTodo = {onDeleteTodo} onCheckTodo= {onCheckTodo} />
      ))}
    </ul>
  );
}

export default TodoList;
