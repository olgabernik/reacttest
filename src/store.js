import { configureStore } from '@reduxjs/toolkit';
import todosReducer from './Slices/todosSlice';
import filtersReducer from './Slices/filtersSlice';
import { loadState } from './localStorageRepository';
import localStorageMiddleware from './localStorageMiddleware';

const persistedState = loadState();

const store = configureStore({
  reducer: {
    todoitems: todosReducer,
    todofilter: filtersReducer,
  },
  middleware: [localStorageMiddleware],
  preloadedState: persistedState,
});

export default store;
