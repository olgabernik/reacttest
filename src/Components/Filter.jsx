import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';
import Text from './Text';

import { FILTER_ALL, FILTER_ACTIVE, FILTER_COMPLETED } from '../constants';

import styles from '../Modules/Filter.module.scss';

const filters = [FILTER_ALL, FILTER_ACTIVE, FILTER_COMPLETED];

function Filter({
  count, onSelectFilter, onClearCompleted, selectedFilter,
}) {
  return (
    <div className={styles.filter}>
      <Text text={`${count} items left`} />
      <span className={styles.divider} />
      <div className={styles.buttons}>
        {filters.map((filter) => (
          <Button
            name={filter}
            key={filter}
            onClick={onSelectFilter}
            parameter={filter}
            isSelected={filter === selectedFilter}
          />
        ))}
      </div>
      <span className={styles.divider} />
      <Button name="Clear Completed" onClick={onClearCompleted} />
    </div>
  );
}

Filter.propTypes = {
  count: PropTypes.number.isRequired,
  onSelectFilter: PropTypes.func.isRequired,
  onClearCompleted: PropTypes.func.isRequired,
  selectedFilter: PropTypes.string.isRequired,
};

export default Filter;
