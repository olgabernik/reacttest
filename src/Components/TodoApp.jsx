import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { todoActions } from '../Slices/todosSlice';
import { setFilter } from '../Slices/filtersSlice';
import filteredTodosSelector from '../filterSelectors';
import {
  ADD_TODO,
  CHECK_TODO,
  CLEAR_COMPLETED,
  DELETE_TODO,
  FILTER_ALL,
  REORDER_TODOS,
} from '../constants';

import TodoList from './TodoList';
import Input from './Input';
import Filter from './Filter';

import styles from '../Modules/TodoApp.module.scss';

function TodoApp() {
  const dispatch = useDispatch();
  let todos = useSelector((state) => state.todoitems);
  const count = todos ? todos.filter((todo) => !todo.isCompleted).length : 0;
  const filteredTodos = useSelector(filteredTodosSelector);
  todos = filteredTodos || todos;

  const [inputValue, setInputValue] = useState('');

  const [selectedFilter, setSelectedFilter] = useState(FILTER_ALL);

  const handleInputChange = (event) => {
    setInputValue(event.target.value);
  };

  const handleAddTodo = (event) => {
    event.preventDefault();
    if (inputValue.trim() !== '') {
      const newItem = { id: uuidv4(), text: inputValue, isCompleted: false };
      dispatch(todoActions[ADD_TODO](newItem));
      setInputValue('');
    }
  };

  const handleDeleteTodo = (id) => {
    // eslint-disable-next-line no-alert
    const confirmed = window.confirm(
      'Are you sure you want to delete this todo?',
    );
    if (confirmed) {
      dispatch(todoActions[DELETE_TODO](id));
    }
  };

  const handleCheckTodo = (id) => {
    dispatch(todoActions[CHECK_TODO](id));
  };

  const onSelectFilter = (filter = FILTER_ALL) => {
    setSelectedFilter(filter);
    dispatch(setFilter(filter));
  };

  const onClearCompleted = () => {
    // eslint-disable-next-line no-alert
    const confirmed = window.confirm(
      'Are you sure you want to clear all completed todos?',
    );
    if (confirmed) {
      dispatch(todoActions[CLEAR_COMPLETED]());
    }
  };

  const handleDragEnd = (result) => {
    if (!result.destination) return;
    dispatch(todoActions[REORDER_TODOS](result));
  };

  return (
    <>
      <span className={styles.header}>TO DO</span>
      <form onSubmit={handleAddTodo}>
        <Input inputValue={inputValue} handleInputChange={handleInputChange} placeholder="Please enter text here..." maxLength="50" />
      </form>
      <div className={styles.container}>
        {todos.length === 0 ? (
          <span className={styles.text}>No todos found...</span>
        ) : (
          <TodoList
            todos={todos}
            onDeleteTodo={handleDeleteTodo}
            onCheckTodo={handleCheckTodo}
            onDragEnd={handleDragEnd}
          />
        )}
        <Filter
          count={count}
          onSelectFilter={onSelectFilter}
          onClearCompleted={onClearCompleted}
          selectedFilter={selectedFilter}
        />
      </div>

    </>
  );
}

export default TodoApp;
