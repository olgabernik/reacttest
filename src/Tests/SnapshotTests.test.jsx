import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import TodoList from '../Components/TodoList';
import Button from '../Components/Button';
import Checkbox from '../Components/Checkbox';
import Filter from '../Components/Filter';
import Input from '../Components/Input';
import TodoApp from '../Components/TodoApp';
import store from '../store';

describe('TodoList', () => {
  const onDeleteTodo = jest.fn();
  const onCheckTodo = jest.fn();
  const onDragEnd = jest.fn();

  it('TodoList should match the snapshot', () => {
    const todos = [
      { id: '1', text: 'Todo 1', isCompleted: false },
      { id: '2', text: 'Todo 2', isCompleted: true },
    ];

    const { asFragment } = render(
      <TodoList
        todos={todos}
        onDeleteTodo={onDeleteTodo}
        onCheckTodo={onCheckTodo}
        onDragEnd={onDragEnd}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Button should match the snapshot', () => {
    const onClick = jest.fn();
    const isSelected = true;

    const { asFragment } = render(
      <Button
        name="Test Button"
        onClick={onClick}
        parameter="testParameter"
        isSelected={isSelected}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Checkbox should match the snapshot', () => {
    const isCompleted = true;

    const { asFragment } = render(
      <Checkbox
        isCompleted={isCompleted}
        onCheckTodo={onCheckTodo}
        todoid="1"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Filter should match the snapshot', () => {
    const onSelectFilter = jest.fn();
    const onClearCompleted = jest.fn();

    const { asFragment } = render(
      <Filter
        count={5}
        onSelectFilter={onSelectFilter}
        onClearCompleted={onClearCompleted}
        selectedFilter="all"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Input should match the snapshot', () => {
    const inputValue = 'Example input';
    const handleInputChange = jest.fn();

    const { asFragment } = render(
      <Input inputValue={inputValue} handleInputChange={handleInputChange} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('TodoApp should match the snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <TodoApp />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
