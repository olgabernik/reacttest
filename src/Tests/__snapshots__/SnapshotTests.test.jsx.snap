// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Button should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="buttonClicked"
    type="button"
  >
    Test Button
  </button>
</DocumentFragment>
`;

exports[`TodoList Checkbox should match the snapshot 1`] = `
<DocumentFragment>
  <input
    checked=""
    class="roundCheckbox"
    type="checkbox"
  />
</DocumentFragment>
`;

exports[`TodoList Filter should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="filter"
  >
    <span
      class="text"
    >
      5 items left
    </span>
    <span
      class="divider"
    />
    <div
      class="buttons"
    >
      <button
        class="button"
        type="button"
      >
        All
      </button>
      <button
        class="button"
        type="button"
      >
        Active
      </button>
      <button
        class="button"
        type="button"
      >
        Completed
      </button>
    </div>
    <span
      class="divider"
    />
    <button
      class="button"
      type="button"
    >
      Clear Completed
    </button>
  </div>
</DocumentFragment>
`;

exports[`TodoList Input should match the snapshot 1`] = `
<DocumentFragment>
  <input
    class="input"
    data-testid="input-new-todo"
    maxlength="50"
    placeholder="some text"
    type="text"
    value="Example input"
  />
</DocumentFragment>
`;

exports[`TodoList TodoApp should match the snapshot 1`] = `
<DocumentFragment>
  <span
    class="header"
  >
    TO DO
  </span>
  <form>
    <input
      class="input"
      data-testid="input-new-todo"
      maxlength="50"
      placeholder="Please enter text here..."
      type="text"
      value=""
    />
  </form>
  <div
    class="container"
  >
    <span
      class="text"
    >
      No todos found...
    </span>
    <div
      class="filter"
    >
      <span
        class="text"
      >
        0 items left
      </span>
      <span
        class="divider"
      />
      <div
        class="buttons"
      >
        <button
          class="buttonClicked"
          type="button"
        >
          All
        </button>
        <button
          class="button"
          type="button"
        >
          Active
        </button>
        <button
          class="button"
          type="button"
        >
          Completed
        </button>
      </div>
      <span
        class="divider"
      />
      <button
        class="button"
        type="button"
      >
        Clear Completed
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TodoList TodoList should match the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="todoList"
    data-testid="todo-list"
  >
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="todo"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="1"
      data-rbd-draggable-context-id="0"
      data-rbd-draggable-id="1"
      data-testid="todo-item"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <input
        class="roundCheckbox"
        type="checkbox"
      />
      <span
        class="text"
      >
        Todo 1
      </span>
      <button
        class="button"
        type="button"
      >
        X
      </button>
    </li>
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="todo completed"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="2"
      data-rbd-draggable-context-id="0"
      data-rbd-draggable-id="2"
      data-testid="todo-item"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <input
        checked=""
        class="roundCheckbox"
        type="checkbox"
      />
      <span
        class="text"
      >
        Todo 2
      </span>
      <button
        class="button"
        type="button"
      >
        X
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
